File: solutions/ocaml/133-binary-search.ml
(*
  Problem 133: Binary Search (LeetCode 704)
  Difficulty: Easy
  Language: OCaml
*)
let binary_search nums target =
  let arr = Array.of_list nums in
  let rec search lo hi =
    if lo > hi then -1
    else
      let mid = lo + (hi - lo) / 2 in
      if arr.(mid) = target then mid
      else if arr.(mid) < target then search (mid + 1) hi
      else search lo (mid - 1)
  in search 0 (Array.length arr - 1)

File: solutions/haskell/133-binary-search.hs
{-
  Problem 133: Binary Search (LeetCode 704)
  Difficulty: Easy
  Language: Haskell
-}
binarySearch :: [Int] -> Int -> Int
binarySearch nums target = search 0 (length nums - 1)
  where
    arr = listArray (0, length nums - 1) nums
    search lo hi
      | lo > hi        = -1
      | arr ! mid == target = mid
      | arr ! mid < target  = search (mid + 1) hi
      | otherwise           = search lo (mid - 1)
      where mid = lo + div (hi - lo) 2
